File: bench/affine_mix.c
// benchmark-contract: affine-mix-xs9;seed=123456789;iterations=20000000;mask=0x03ffffffffffffff;xorshift=9
#include <inttypes.h>
#include <stdint.h>
#include <stdio.h>

static inline void emit_checksum(uint64_t value) {
  // Every peer prints this one line and nothing else; the 63-bit mask
  // keeps the value printable by the languages without unsigned 64-bit
  // integers (NURL's `i`, Python's signed int, JS BigInt) so the five
  // outputs can be compared byte for byte.
  printf("%llu\n", (unsigned long long)(value & 0x7fffffffffffffffULL));
}

int main(void) {
  const uint64_t iterations = 20000000ULL;
  uint64_t state = 123456789ULL;
  uint64_t i = 0;

  while (i < iterations) {
    state = ((state << 3) + i) & 0x03ffffffffffffffULL;
    state ^= state >> 9;
    state = ((state << 2) - 3ULL) & 0x03ffffffffffffffULL;
    ++i;
  }

  emit_checksum(state);
  return 0;
}

File: bench/affine_mix.rs
// benchmark-contract: affine-mix;seed=123456789;iterations=50000000;mask=0x03ffffffffffffff
const ITERATIONS: u64 = 50_000_000;
const SEED: u64 = 123_456_789;
const MASK: u64 = 0x03ff_ffff_ffff_ffff;

fn finish(value: u64) -> ! {
    #[cfg(bench_verify)]
    std::io::Write::write_all(&mut std::io::stdout(), &value.to_le_bytes()).unwrap();
    std::process::exit((value & 0x7f) as i32)
}

fn main() {
    let mut state = SEED;
    let mut i = 0u64;
    while i < ITERATIONS {
        state = (state << 3).wrapping_add(i) & MASK;
        state = (state << 2).wrapping_sub(3) & MASK;
        i += 1;
    }
    finish(state);
}
